test_unstructured/partition/common/test_metadata.py

```diff
@@ -62,73 +62,139 @@ def but_it_returns_None_when_there_is_no_file_at_that_path(self, tmp_path: pathl
-    assert elements[4].metadata.parent_id == elements[3].id, "ListItem should be child of Title"
-    assert elements[5].metadata.parent_id == elements[3].id, "ListItem should be child of Title"
-    assert elements[6].metadata.parent_id == elements[0].id, "ListItem should be child of Title"
-    # NOTE(Hubert): moving the category field to Element, caused this to fail.
-    # Checkboxes will soon be deprecated, then we can remove the test.
-    # assert (
-    #         elements[7].metadata.parent_id is None
-    # ), "CheckBox should be None, as it's not a Text based element"
-    assert elements[8].metadata.parent_id is None, "Title 2 should be child of None"
-    assert elements[9].metadata.parent_id == elements[8].id, "ListItem should be child of Title 2"
-    assert elements[10].metadata.parent_id == elements[8].id, "ListItem should be child of Title 2"
-    assert elements[11].metadata.parent_id == elements[8].id, "Text should be child of Title 2"
-
-
-def test_set_element_hierarchy_custom_rule_set():
-    elements_to_set = [
-        Header(text="Header"),  # 0
-        Title(text="Title"),  # 1
-        NarrativeText(text="NarrativeText"),  # 2
-        Text(text="Text"),  # 3
-        Title(text="Title 2"),  # 4
-        FigureCaption(text="FigureCaption"),  # 5
-    ]
+class Describe_set_element_hierarchy:
+
+    def it_applies_default_ruleset(self):
+        elements = [
+            Title(element_id="0", text="Title0"),
+            Text(element_id="1", text="Text0"),
+            Header(element_id="2", text="Header0"),
+            Text(element_id="3", text="Text1"),
+            Title(element_id="4", text="Title1"),
+            Text(element_id="5", text="Text2"),
+        ]
+
+        result = set_element_hierarchy(elements)
+
+        assert result[0].metadata.parent_id is None
+        assert result[1].metadata.parent_id == "0"  # Text0 is under Title0
+        assert result[2].metadata.parent_id is None  # Header0 is higher than Title0
+        assert result[3].metadata.parent_id == "2"  # Text1 is under Header0
+        assert result[4].metadata.parent_id == "2"  # Title1 is under Header0
+        assert result[5].metadata.parent_id == "4"  # Text2 is under Title1, which is under Header0
+
+    def it_applies_ruleset_based_on_element_category_regardless_of_depth(self):
+        elements = [
+            Title(element_id="0", text="Title", metadata=ElementMetadata(category_depth=2)),
+            Text(element_id="1", text="Text", metadata=ElementMetadata(category_depth=1)),
+            Header(element_id="2", text="Header", metadata=ElementMetadata(category_depth=2)),
+            Text(element_id="3", text="Text", metadata=ElementMetadata(category_depth=1)),
+        ]
+
+        result = set_element_hierarchy(elements)
+
+        assert result[1].metadata.parent_id == "0"  # Text is under Title despite category_depth
+        assert result[3].metadata.parent_id == "2"  # Text is under Header despite category_depth
```

Comment:
Maybe the name can more clearly describe the behavior. Are we saying something like `it_ignores_category_depth_except_on_top_level_elements` or something like that?

A separate behavior might be "it_does_not_mix_hierarchies_between_rulesets" or something. Like if you had `Header cat_lvl=1` (which actually wouldn't make any sense since headers don't have levels like section-headings do) followed by `Title cat_lvl=0`, would the `Title` use the `Header` as its parent or become a top-level (parentless) element?

Maybe you could describe the behavior you're looking to verify here in as many words as it takes, then we could look for a short way to say it. The docstring can also be used to elaborate on the behavior.

Reply:
So, your example is what I'm trying to illustrate. I was expecting the `category_level` to supersede the element `category`, but it doesn't, it's ignored. The `category_level` is ignored when it comes to comparing different _types_ of elements. In the test, I set the `category_depth` of the `Title` elements to be "under" the `Text` elements (setting them to `"2"`), but that is ignored and the only important factor is the element's `category` (`Title`/`Text`).

Reply:
I feel like the semantics are fine, but the vocabulary might be off--I was trying to reuse the terms used in the file: "ruleset", "category", "category_depth" etc. It's a little confusing to have "category" for the element type and have "category_depth"

Reply:
Oooh, okay, that _is_ weird!

Okay, I've had a look at the code and I don't think I can fully understand it without incrementally executing a bunch of cases. If I did take the time to do that I'm sure it would result in me wanting to make changes.

So I'm thinking time to cut bait unless there's a clear problem we're trying to diagnose :)

I'm fine with this one just the way it is :)

Reply:
I just realized we have `category_depth` kind of doing double-duty. It's used internally to determine parent, but we also make it accessible to users--I have to factor that into my partitioning of onenote...

Reply:
Okay, I should have known I wouldn't be able to help myself, so after a further look, here are some possible insights:

Q. What is a _rule-set_?
A. A rule-set is is a `{parent: list[permissible_children]}` mapping, where both parent and permitted-child are expressed as an `element.category` (str/name) value.

Inferences:
1. Only an element-type that appears as a key (parent) in the rule-set mapping can be the target of a `.metadata.parent_id` reference.

Loop behavior: for each element:
- the current "parent" is the top element on the stack.
- the current element gets the current parent-id when the current element meets either of these criteria:
    - it is the same kind as the parent but a lower level, like both are `Title` but the current element is level 2 instead of 1
    - it's not the same kind but the current element is one of the permissible children of the current parent
- Otherwise, remove parents from the stack one at a time until you either:
    - find a suitable parent, or
    - run out of parents on the stack

In the later case (run out of parents) the the element gets `parent_id=None`.

There are more subtleties this doesn't cover, like how does the first parent get there and so on, but I'm already feeling like rewriting it, so I'd better stop here :)